Yoga build host image on arc

Alex-Welsh opened the pull request:
Still putting the final touches to this PR but everything should be working correctly now.
All images built [here](https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/8433465635)
latest test [here](https://github.com/stackhpc/stackhpc-kayobe-config/actions/workflows/overcloud-host-image-build.yml)

Alex-Welsh commented:
Current issues:
Rocky 8 images are building as rocky 9
Rocky 9 images are failing to boot due to to an issue with cloud-init config not being parseable:
```
[   12.623671] cloud-init[966]: 2024-03-27 09:48:06,919 - util.py[WARNING]: Failed loading yaml blob. Invalid format at line 102 column 5: "while parsing a block mapping
[   12.626039] cloud-init[966]:   in "<unicode string>", line 102, column 5:
[   12.627229] cloud-init[966]:         renderers: ['network-manager']
[   12.628358] cloud-init[966]:         ^
[   12.629054] cloud-init[966]: expected <block end>, but found '<block mapping start>'
[   12.630460] cloud-init[966]:   in "<unicode string>", line 103, column 7:
[   12.631628] cloud-init[966]:           activators: ['network-manager']
```
I think that issue is caused by a whitespace error on [this line](https://github.com/stackhpc/stackhpc-image-elements/blob/841947d98ca2dedf3b3b8a463f23d5507ccbcb06/elements/rocky-container-stackhpc/containerfiles/9-stackhpc#L9C110-L9C111) of the containerfile
A fix is [in progress](https://github.com/stackhpc/stackhpc-image-elements/compare/master...fix-rocky-containerfile)

markgoddard commented on terraform/host-image-builder/README.rst:
```diff
@@ -0,0 +1,9 @@
+============================
```
Was it not possible to reuse the aio TF config?

Alex-Welsh replied:
I have a feeling there used to be a conflict in an earlier iteration that doesn't exist any longer. I'll have a go at merging them.

Alex-Welsh replied:
Seems fine now

markgoddard commented on terraform/host-image-builder/vm.tf:
```diff
@@ -0,0 +1,98 @@
+
+resource "openstack_compute_instance_v2" "kayobe-host-image-builder" {
+  name         = var.host_image_builder_name
+  flavor_name  = "en1.medium"
```
host_image_builder_flavor

Alex-Welsh replied:
Removed, using AIO

markgoddard commented on etc/kayobe/pulp-repo-versions.yml:
```diff
@@ -64,8 +64,8 @@ stackhpc_pulp_repo_rocky_9_3_crb_version: 20231215T005810
-stackhpc_pulp_repo_ubuntu_focal_version: 20231018T165217
-stackhpc_pulp_repo_ubuntu_jammy_security_version: 20231020T074329
-stackhpc_pulp_repo_ubuntu_jammy_version: 20231020T074329
+stackhpc_pulp_repo_ubuntu_cloud_archive_version: 20240327T041648
```
We might want to control when we roll this out separately. I guess we don't necessarily need/want to update the images in this PR.

Alex-Welsh replied:
Good point. I'll leave it in for now since CI is useful for determining whether I'm building valid images, but remove it later

Alex-Welsh replied:
reverted

markgoddard commented on etc/kayobe/ansible/openstack-host-image-upload.yml:
```diff
@@ -0,0 +1,42 @@
+---
+# This playbook is desinged to be used by the overcloud-host-image-build.yml
```
Have you considered using the Grammarly VS Code plugin?
```suggestion
# This playbook is designed to be used by the overcloud-host-image-build.yml
```

Alex-Welsh replied:
Done <sup><sup>with great shame</sup></sup>

markgoddard commented on etc/kayobe/ansible/openstack-host-image-upload.yml:
```diff
@@ -0,0 +1,42 @@
+    - name: Write out secure.yml
+      no_log: true
+      vars:
+        - os_secrets:
```
nit: no need for a list here (although it does work)

Alex-Welsh replied:
Feels wrong but I'll go with it

markgoddard commented on etc/kayobe/ansible/openstack-host-image-upload.yml:
```diff
@@ -0,0 +1,42 @@
+                  application_credential_secret: "{{ lookup('ansible.builtin.env', 'OS_APPLICATION_CREDENTIAL_SECRET') }}"
+      copy:
+        content: "{{ os_secrets | to_nice_yaml }}"
+        dest: secure.yaml
```
It might be worth removing this guy at the end using a block with `always`, and possibly clouds.yaml too.

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/openstack-host-image-upload.yml:
```diff
@@ -0,0 +1,42 @@
+        dest: clouds.yaml
+        mode: 0664
+
+    - name: Write out secure.yml
```
```suggestion
    - name: Write out secure.yaml
```

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/openstack-host-image-upload.yml:
```diff
@@ -0,0 +1,42 @@
+    local_image_path: "/opt/kayobe/images/overcloud-{{ os_distribution }}-{{ os_release }}/overcloud-{{ os_distribution }}-{{ os_release }}.qcow2"
+    image_name: "overcloud-{{ os_distribution }}-{{ os_release }}"
+  tasks:
+    - name: Write out clouds.yml
```
```suggestion
    - name: Write out clouds.yaml
```

Alex-Welsh replied:
Done

markgoddard commented on etc/kayobe/ansible/openstack-host-image-upload.yml:
```diff
@@ -0,0 +1,42 @@
+      copy:
+        content: "{{ os_secrets | to_nice_yaml }}"
+        dest: secure.yaml
+        mode: 0664
```
0600 or 0660

Alex-Welsh replied:
Done

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -35,21 +35,39 @@ on:
+
+      - name: Start the SSH service
+        run: |
+          sudo /etc/init.d/ssh start
```
Why is this required?

Alex-Welsh replied:
I just couldn't get to SSH to the remote machine without it. I can dig out the error if you're interested

markgoddard replied:
Would definitely like to know why.

Alex-Welsh replied:
https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/8467978144/job/23199884304#step:22:216
ssh-keyscan just failed to run

markgoddard replied:
Ah. Looks like we don't do a control host bootstrap in the container build job, which is why we didn't hit this. And the aio job uses kayobe-automation which does it's own SSH stuff

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -110,36 +96,133 @@ jobs:
+
+      - name: Update inventory
+        run: |
+          cat > src/kayobe-config/etc/kayobe/environments/ci-builder/inventory/hosts << EOF
```
Can we change this file to look like this out of the box instead?

Alex-Welsh replied:
My assumption was that it could interfere with container image builds. I'll kick one off and see what happens

markgoddard replied:
We write out the inventory in the container build job, so shouldn't cause a problem.

Alex-Welsh replied:
Done

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -110,36 +96,133 @@ jobs:
 
-      - name: Configure the seed host
+      - name: Configure the seed host (Builder VM)
+        continue-on-error: true
```
Why?

Alex-Welsh replied:
Old debugging artifact

Alex-Welsh replied:
Removed

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -110,36 +96,133 @@ jobs:
+          source venvs/kayobe/bin/activate &&
+          source src/kayobe-config/kayobe-env --environment ci-builder &&
+          kayobe seed host command run \
+          --command "sudo dnf config-manager --set-enabled crb && sudo dnf -y install epel-release && sudo dnf -y install zstd debootstrap kpartx cloud-init" --show-output
```
What's crb for?

markgoddard replied:
Surprised we don't have cloud-init

Alex-Welsh replied:
CRB is a prerequisite for EPEL https://wiki.rockylinux.org/rocky/repo/#notes-on-epel
I might be able to reduce the requirements here but it would take a bit of time to test.

markgoddard replied:
Probably not worth the time

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -110,36 +96,133 @@ jobs:
+          kayobe seed host configure -e seed_bootstrap_user=rocky --skip-tags network
+
+      - name: Install dependencies
+        continue-on-error: true
```
Why?

Alex-Welsh replied:
Old debugging artifact

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -243,6 +342,16 @@ jobs:
           KAYOBE_VAULT_PASSWORD: ${{ secrets.KAYOBE_VAULT_PASSWORD }}
         if: inputs.rocky9
 
+      - name: Show last error logs
```
Is this just for testing?

Alex-Welsh replied:
I initially added it for testing before I could copy back the full logs, but I was going to leave it in since it's very convenient to just check for obvious errors.

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -345,83 +470,27 @@ jobs:
+        continue-on-error: true
+        run: |
+          mkdir logs
+          scp -r rocky@$(jq -r .access_ip_v4.value src/kayobe-config/etc/kayobe/environments/ci-builder/tf-outputs.yml):/opt/kayobe/images/*/*.std* ./logs/
```
Does this need to specify the SSH key?

Alex-Welsh replied:
Nope, seems to work without

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -110,36 +96,141 @@ jobs:
+          ssh_username = "rocky"
+          aio_vm_name = "skc-host-image-builder"
+          # Must be a Rocky Linux 9 host to successfully build all images
+          # This MUST NOT be an LVM image. It can cause confusing conficts with the built image.
```
heh, I remember this one. It happens if the VG has the same name on the host. Tried to fix but was unable to.

Alex-Welsh replied:
Drove me insane for a bit that one

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -110,36 +96,141 @@ jobs:
+          EOF
+        working-directory: ${{ github.workspace }}/src/kayobe-config/terraform/aio
+        env:
+          SSH_USERNAME: "id_rsa.pub"
```
Not a username

Alex-Welsh replied:
removed

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -110,36 +96,141 @@ jobs:
+        working-directory: ${{ github.workspace }}/src/kayobe-config/terraform/aio
+        env:
+          SSH_USERNAME: "id_rsa.pub"
+          VM_NAME: "skc-host-image-builder-${{ github.run_id }}"
+          VM_IMAGE: 
+          VM_FLAVOR: 
+          VM_NETWORK: 
+          VM_SUBNET: 
+          VM_INTERFACE: ${{ inputs.vm_interface }}
```
Unused

Alex-Welsh replied:
removed

markgoddard commented on .github/workflows/overcloud-host-image-build.yml:
```diff
@@ -110,36 +96,141 @@ jobs:
+          VM_NETWORK: 
+          VM_SUBNET: 
+          VM_INTERFACE: ${{ inputs.vm_interface }}
+          VM_TAGS: '["skc-host-image-build", "ref=${{ github.ref_name }}"]'
```
Unused

Alex-Welsh replied:
Removed

markgoddard commented on etc/kayobe/stackhpc-overcloud-dib.yml:
```diff
@@ -21,7 +21,7 @@ stackhpc_overcloud_dib_name: "deployment_image"
 
 # StackHPC overcloud DIB image elements.
 stackhpc_overcloud_dib_elements:
-  - "{{ os_distribution }}-{% if os_distribution == 'rocky' %}container-stackhpc{% else %}minimal{% endif %}"
```
I'd consider this as a separate PR to fix RL9 builds

Alex-Welsh replied:
Agreed. It's also not fully fixed yet and I don't want to sink more time into it now so I'll put that on hold.

Alex-Welsh committed:
Fix host image builds on Arc runners

Arc runners are kubernetes-orchestrated github runners. Host image
builds do not work on these runners, so this commit adapts the host
image build workflow to spin up a worker VM which executes the build.

markgoddard commented on etc/kayobe/overcloud-dib.yml:
```diff
@@ -71,7 +71,7 @@ overcloud_dib_host_packages_extra:
   - repo: "https://github.com/stackhpc/stackhpc-image-elements"
     local: "{{ source_checkout_path }}/stackhpc-image-elements"
-    version: "v1.6.0"
+    version: "v1.6.1"
```
Should this be in a separate change?

Alex-Welsh replied:
Builds flat out fail without it, so I'd argue it should remain part of this change

markgoddard replied:
Ack
